### PR TITLE
Table connecting verbs and modes

### PR DESCRIPTION
**Pull request by michielbdejong**
See https://github.com/michielbdejong/web-access-control-spec/blob/verbs-modes-table/README.md#relation-between-http-verbs-and-modes-of-access

### CODE REVIEW

**Comment by TallTed** [comment]
> ties things closely to HTTP

Perhaps you might be less against the idea if surrounding phrasing described the HTTP verbs as clarifying examples, rather than implying (if only by not mentioning any other verbs/systems) that they're the only possible methods?

**Comment by michielbdejong** [comment]
Let's discuss this in person on Monday!

**Comment by michielbdejong** [comment]
Sure, we can move this to a note, or even a wiki page, or part of the readme of the testsuite, as long as it's written down somewhere exactly what modes are required for exactly which actions.

**Comment by michielbdejong** [comment]
@RubenVerborgh @acoburn changes applied, please re-review?

File: README.md [hunk 1]
```diff
@@ -489,12 +489,12 @@ under which circumstances:
...
-| PUT | /foo/bar | /foo/ exists | Append or Write on /foo/ + Write on resulting resource |
-| PUT | /foo/bar | /foo/ gets created | Append or Write on / + Append or Write on /foo/ + Write on /foo/bar |
-| PATCH | /foo/bar | patch instructions contain deletions | Write on /foo/bar |
+| PUT | /foo/bar | /foo/ exists, will contain /foo/bar | Append or Write on /foo/ + Write on resulting resource |
+| PUT | /foo/bar | /foo/ gets created, will contain /foo/bar | Append or Write on / + Append or Write on /foo/ + Write on /foo/bar |
+| PATCH | /foo/bar | patch instructions contain deletions | Read and Write on /foo/bar |
 | PATCH | /foo/bar | patch instructions contain only insertions | Append or Write on /foo/bar |
 | DELETE | /foo/bar | | Write on /foo/bar |
-| DELETE | /foo/ | /foo/ is empty | Write on /foo/ |
+| DELETE | /foo/ | No resources other than /foo/ will get deleted | Write on /foo/ |
```
Comment (acoburn) [inline README.md]: How is this different than the `DELETE` note in the line above? Perhaps just adding the note about "No resource other than..." to the delete method above

File: README.md [hunk 2]
```diff
@@ -489,12 +489,12 @@ under which circumstances:
 | HEAD |  |   |  Read |
 | GET |  |   |  Read |
 | POST | /foo/ | /foo/ exists | Append or Write on /foo/ + Write on resulting resource |
-| PUT | /foo/bar | /foo/ exists | Append or Write on /foo/ + Write on resulting resource |
-| PUT | /foo/bar | /foo/ gets created | Append or Write on / + Append or Write on /foo/ + Write on /foo/bar |
-| PATCH | /foo/bar | patch instructions contain deletions | Write on /foo/bar |
+| PUT | /foo/bar | /foo/ exists, will contain /foo/bar | Append or Write on /foo/ + Write on resulting resource |
+| PUT | /foo/bar | /foo/ gets created, will contain /foo/bar | Append or Write on / + Append or Write on /foo/ + Write on /foo/bar |
```
Comment (acoburn) [inline README.md]: This interaction (recursive create of intermediate directories) is not defined in the specification. If we are describing this as an example, it should be clear whether this is an expected interaction.

File: README.md [hunk 3]
```diff
@@ -489,12 +489,12 @@ under which circumstances:
...
+| PATCH | /foo/bar | patch instructions contain deletions | Read and Write on /foo/bar |
 | PATCH | /foo/bar | patch instructions contain only insertions | Append or Write on /foo/bar |
```
Comment (acoburn) [inline README.md]: For `PATCH` instructions that only contain insertions, I can imagine that `Read` would be useful or necessary in an implementation. For example, given the document:

```turtle
<> a <https://example.com/MyType> .
```
the SPARQL Query:

```sparql
INSERT {
    <> a <https://example.com/MyType> .
} WHERE {}
```

would result in no change to the original Graph, but that cannot be determined unless the document is first read into a Graph.

File: README.md (hunk 1 above)
Comment (TallTed) [reply README.md]: `DELETE /foo/bar` seems obviously different than `DELETE /foo/`... 

The first (tries to) `DELETE` a resource (`/foo/bar`).
The second (tries to) `DELETE` a container (`/foo/`).

That said, in several (but not all) other rows, the comment in the 3rd column is read with the titular preceding "if...", so "No resources other than /foo/ will get deleted" is definitely incorrect, and `/foo/ is empty` was closer to correct.  Possibly, `[if] /foo/ contains no other resources` would be better.  Possibly, the latter columns of this table should be reconsidered and rewritten overall, perhaps adding a column for "results" such that multiple `IF` conditions (e.g., `if /foo/ contains another resource`) could also be addressed in this table.

File: README.md (hunk 2 above)
Comment (michielbdejong) [reply README.md]: True! This implicitly ties WAC to Solid which we don't want. We should say 'For instance like when you're using WAC as part of Solid which has `mkdir -p` behaviour', and then link to https://github.com/solid/solid-spec/blob/master/api-rest.md#http-put-to-create

File: README.md (hunk 1 above)
Comment (michielbdejong) [reply README.md]: All true! The root problem here is that the 'container is empty' thing implicitly ties WAC to Solid which we don't want. We should say 'For instance like when you're using WAC as part of Solid', and then link to https://github.com/solid/solid-spec/blob/master/api-rest.md#deleting-containers

**Comment by michielbdejong** [comment]
The data browser Sharing pane offers 5 useful defaults:
* 'owners' is read+write+control, "can read, write, and control sharing."
* 'editors' is read+write, "can read and change information"
* 'posters' is read+append, "can add new information, and read but not change existing information"
* 'submitters' is append, "can add new information but not read any"
* 'viewers' is read, "can read but not change information"

This is then to be interpreted differently for a container than for a non-container. For a container, read permissions means you can read its `ldp:contains` triples. For a non-container resource, it means you can read its contents as it was last written or updated through POST, PUT or PATCH (and optionally make the server do some RDF operations on it).

For creating a new resource, I confirmed empirically against NSS that if you choose the URL with PUT, then write on the container is not enough, you also need write on the resulting resource. Append on the resulting resource is also not enough then.

But what's funny is that if you append to a container with POST, then append on the container is enough and no rights at all at the resulting resource are needed.

For DELETE, write on the container seems to be enough. You don't need to have any write permissions on the resource. This is weird because it means that there can be resources which you can DELETE but not PUT or PATCH.

**Comment by michielbdejong** [comment]
As mentioned in https://github.com/solid/solid-spec/pull/193#issuecomment-574554383 the success/failure of a DELETE can expose existence, and therefore it should require that read access. It's a diversion from POSIX but it makes sense and it's how NSS currently implemented it for PATCH. That discusssion is only in the context of deleting triples from an RDF source, but we *could* equally apply it to containers; to delete a resource you should probably have access to read the container it's in.
However, that's not what NSS does (I just tried:
```sh
curl -X DELETE https://michielbdejong.inrupt.net/public/public-write/test.ttl
```
and that succeeded without read access to 
curl -X DELETE https://michielbdejong.inrupt.net/public/public-write/)
so I guess the reasoning for deleting triples inside a file is slightly different from the one applied for deleting files from a folder. So this table is correct in stating that DELETE does not require container read, but PATCH containing deletions does require Read in addition to Write.

**Pull request closed without merging**